common/src/chain/block/consensus_data.rs

```diff
@@ -0,0 +1,44 @@
+use crate::chain::TxOutput;
+use crate::primitives::Compact;
+use parity_scale_codec::{Decode, Encode};
+
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode)]
+pub enum ConsensusData {
+    #[codec(index = 0)]
+    None,
+    #[codec(index = 1)]
+    PoW(PoWData),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode)]
+pub struct PoWData {
+    bits: Compact,
```

Comment:
I know this would deviate from Bitcoin but the difficulty is determined by the time difference of the last two blocks that are at a height that is a multiple of 2016, so this field is not needed. Even if the field is included, we need to check it has the correct value if it's relied upon for the difficulty check. Have you thought about getting rid of it?

Reply:
The bits field is being used in the ff. ways:
1. When `no_retargeting` field in the config is set to `true`. That only happens for Regtest. https://github.com/mintlayer/mintlayer-core/blob/1e390f95555ed9a29c696174fd2f3781c6333761/common/src/chain/pow.rs#L59
2. if it's not due for retargeting yet, and `allow_min_dfficulty_blocks` is set to `false`, which is the Mainnet and Signet: https://github.com/mintlayer/mintlayer-core/blob/1e390f95555ed9a29c696174fd2f3781c6333761/common/src/chain/pow.rs#L65

Let's say the current block height is 8067, and it's not due for retarget (since retarget has happened at 8064). 
I am assuming that the difficulty is retrieved from the previous block, at height 8066. No calculation needed.
If we remove the `bits` field, we'll have to keep calculating the time diff of the blocks of height 8064 and 6048. 

As for checking whether its value is correct, Did you mean the `check_proof_of_work()`? Or do you mean checking whether the previous block's bits field is a correct value?

Reply:
> As for checking whether its value is correct, Did you mean the `check_proof_of_work()`? Or do you mean checking whether the previous block's bits field is a correct value?

What I mean is that the value of the `bits` field cannot be trusted to contain the difficulty anyway. I see the `get_work_required` method here but I can't see it being used anywhere. I.e. at any point, the following must pass:

```rust
assert_eq!(calculate_difficulty_from_the_scratch(...).try_into(), Just(this_block.bits))
```

Blocks that do not satisfy that condition must be rejected, since the `check_proof_of_work` function relies on the `bits` field to contain accurate difficulty.

Reply:
The `get_work_required()`(previously `check_for_work_required()`) will be used in the block production.
This was the earliest implementation : https://github.com/mintlayer/mintlayer-core/pull/62/commits/ed8db7d39d28167eadaef7a97fa9602b277c1113#diff-9a00856cffad13af6904b1cb9d4b5a4fd9e0591f10a35a67ddd805141890eeb6R91

But I removed it in this PR, because it's meant for the block production PR (which is currently closed, until Anton's or this is pushed).

Did you mean because the bits is a u32 representation, while the difficulty is 256, that's why the bits field can't be trusted? Or because the previous block can't be trusted, since the bits field can easily be just any number and not from a calculated value during the generation of the 2016th multiple block? 

I haven't checked in bitcoin where that is, but we can actually add that kind of validation.

I somehow do agree with you; removing the bits field is not so bad, since we can always calculate for the difficulty. Your `get_block_id_by_height` function will be useful (Though I think I have to access it through the `BlockIndex`).

For now I've been following how Bitcoin does it. But also I don't have problems diverging from Bitcoin. 
Although I don't know what Sam @TheQuantumPhysicist thinks, we can wait for his opinion.